src/block_storage/protocol.rs

```diff
@@ -0,0 +1,201 @@
+    enum VolumeSortKey {
+        CreatedAt = "created_at",
+        Id = "id",
+        Name = "name",
+        UpdatedAt = "updated_at"
+    }
+}
+
+impl Default for VolumeSortKey {
+    fn default() -> VolumeSortKey {
+        VolumeSortKey::CreatedAt
+    }
+}
+
+/// A volume attachment.
+#[derive(Debug, Clone, Deserialize)]
+pub struct VolumeAttachment {
+    pub server_id: String, // this should be a reference to a server
+    pub attachment_id: String,
+    pub attached_at: String,
+    pub host_name: String,
+    pub volume_id: String, // this should be a reference to a volume
+    pub device: String,
+    pub id: String,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct Link {
+    pub rel: String,
+    pub href: String,
+}
+
+/// A volume.
+#[derive(Debug, Clone, Deserialize)]
+pub struct Volume {
+    // TODO: not all fields fully match the API spec:
+    // https://docs.openstack.org/api-ref/block-storage/v3/#list-accessible-volumes-with-details
+    // Some fields are not actually optional, but don't work without Option<>.
+    // Others should maybe be enums, but the possible values are not documented.
+    // There are comments for these cases.
+    pub migration_status: Option<String>, // consider enum
+    pub attachments: Vec<VolumeAttachment>,
+    pub links: Vec<Link>,
+    pub availability_zone: Option<String>,
+    #[serde(rename = "os-vol-host-attr:host")]
+    pub host: Option<String>,
+    pub encrypted: bool,
+    pub encryption_key_id: Option<String>,
+    pub updated_at: Option<String>,
+    pub replication_status: Option<String>, // not optional in spec, also consider enum
+    pub snapshot_id: Option<String>,
+    pub id: String,
+    pub size: u64,
+    pub user_id: String,
+    #[serde(rename = "os-vol-tenant-attr:tenant_id")]
+    pub tenant_id: Option<String>,
+    #[serde(rename = "os-vol-mig-status-attr:migstat")]
+    pub migstat: Option<String>, // consider enum
+    pub metadata: HashMap<String, String>,
+    pub status: VolumeStatus,
+    #[serde(rename = "volume_image_metadata")]
+    pub image_metadata: Option<HashMap<String, String>>,
+    pub description: Option<String>,
+    #[serde(rename = "multiattach")]
+    pub multi_attachable: bool,
+    #[serde(rename = "source_volid")]
+    pub source_volume_id: Option<String>,
+    #[serde(rename = "consistencygroup_id")]
+    pub consistency_group_id: Option<String>, // not optional in spec
+    #[serde(rename = "os-vol-mig-status-attr:name_id")]
+    pub name_id: Option<String>,
+    pub name: String,
+    pub bootable: String,
+    pub created_at: String,
+    pub volumes: Option<Vec<Volume>>, // not optional in spec
+    pub volume_type: String,          // consider enum
+    pub volume_type_id: Option<HashMap<String, String>>, // not optional in spec
+    pub group_id: Option<String>,
+    pub volumes_links: Option<Vec<String>>,
+    pub provider_id: Option<String>,
+    pub service_uuid: Option<String>, // not optional in spec
```

Comment:
Maybe rename to service_id? This is the only field that uses _uuid

Reply:
done